NeoMathEngine/src/CPU/CpuMathEngineDnnDropout.cpp

```diff
@@ -40,29 +49,81 @@ void CCpuMathEngine::Dropout( const CDropoutDesc& dropoutDesc, const CFloatHandl
 	const int batchLength = desc.IsBatchwise ? input.ObjectCount() : input.BatchLength();
 	const int batchWidth = input.ObjectCount() / batchLength;
 	const int maskSize = batchWidth * objectSize;
+	
+	CCpuRandom random(desc.seed);
+	CIntArray<maskAlign> generated;
 
-	ASSERT_EXPR( desc.Mask.Size() == maskSize );
+	const int inputObjectSize = input.ObjectSize();
+	const float* inputPointer = GetRaw(inputData);
+	float* outputPointer = GetRaw(outputData);
+	float* mask = GetRaw(desc.Mask.GetHandle());
+	const int cacheSize = desc.Mask.Size();
 
-	if( !desc.IsSpatial ) {
-		MultiplyMatrixByDiagMatrix( inputData, batchLength, maskSize, desc.Mask.GetHandle(),
-			outputData, output.BlobSize() );
-		return;
-	}
+	const float* first;
+	float* result;
 
-	CFloatHandle currInput = inputData;
-	CFloatHandle currOutput = outputData;
-	for( int i = 0; i < input.ObjectCount(); ++i ) {
-		MultiplyMatrixByDiagMatrix( currInput, input.ObjectSize() / objectSize, objectSize,
-			desc.Mask.GetHandle() + ( i % batchWidth ) * objectSize, currOutput, input.ObjectSize() );
-		currInput += input.ObjectSize();
-		currOutput += input.ObjectSize();
-	}
-}
+	if(!desc.IsSpatial) {
 
-CDropoutDesc* CCpuMathEngine::InitDropout( float rate, bool isSpatial, bool isBatchwise,
-	const CBlobDesc& input, const CBlobDesc& output, int seed )
-{
-	return new CMathEngineDropoutDesc( mathEngine(), rate, isSpatial, isBatchwise, input, output, seed );
+		const float* first;
+		float* result;
+		int currSize;
+
+		for(int i = 0; i < (maskSize + cacheSize - 1) / cacheSize; ++i) {
+			currSize = std::min(cacheSize, maskSize - i * cacheSize);
+			int idx = 0;
+
+			first = inputPointer;
+			result = outputPointer;
+
+			const int alignedSize = (currSize + (maskAlign - 1)) / maskAlign;
```

Comment:
Separate the last iteration of this cycle.
For all except last iterations `currSize = cacheSize` and `alignedSize = cacheSize + (maskAlign - 1)) / maskAlign`, both are constants are known, can be stored in `Desc` class.
For last iteration: `currSize =  maskSize - i_last * currSize` and `alignedSize` are also constants, known to be stored in  `Desc`